Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,238 @@
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
+    left_ne_zero_of_max_eq hx H1, div_self, le_refl]
+  · simp only [Fin.isValue, H2, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
+    right_ne_zero_of_max_eq hx H2, div_self, le_refl]
+
+lemma rpow_bound {k : ℝ} (hk : 0 ≤ k) (z : ℍ) (x : Fin 2 → ℤ) (hx : x ≠ 0) :
+    (r z) ^ k * (max (x 0).natAbs (x 1).natAbs) ^ k ≤ Complex.abs (x 0 * z + x 1) ^ k := by
+  rw [← Real.mul_rpow (r_pos _).le (Nat.cast_nonneg _)]
+  refine Real.rpow_le_rpow (mul_nonneg (r_pos _).le (Nat.cast_nonneg _)) ?_ hk
+  let n := max (x 0).natAbs (x 1).natAbs
+  have hn0 : n ≠ 0 := by
+    contrapose! hx
+    rw [Nat.max_eq_zero_iff, Int.natAbs_eq_zero, Int.natAbs_eq_zero] at hx
+    exact funext fun j ↦ by fin_cases j <;> tauto
+  have h11 : x 0 * (z : ℂ) + x 1 = (x 0 / n * z + x 1 / n) * n := by
+    push_cast
+    rw [div_mul_eq_mul_div, ← add_div, div_mul_cancel₀ _ (Nat.cast_ne_zero.mpr hn0)]
+  rw [Nat.cast_max, h11, map_mul, abs_natCast]
+  gcongr
+  · rcases div_max_sq_ge_one x hx with H1 | H2
+    · simpa only [Fin.isValue, ofReal_div, ofReal_intCast, n] using auxbound1 z (x 1 / n) H1
+    · simpa only [Fin.isValue, ofReal_div, ofReal_intCast, n] using auxbound2 z (x 0 / n) H2
+  · simp only [Fin.isValue, Nat.cast_max, le_refl, n]
+
+theorem summand_is_bounded_on_box_rpow {k : ℝ} (hk : 0 ≤ k) (z : ℍ) (n : ℕ) (x : Fin 2 → ℤ)
+    (hx : (x 0, x 1) ∈ box n) : Complex.abs (x 0 * z + x 1) ^ (-k) ≤ (r z) ^ (-k) * n ^ (-k) := by
+  by_cases hn : n = 0
+  · simp only [hn, box_zero, Finset.mem_singleton, Prod.mk_eq_zero] at hx
+    rw [hx.1, hx.2, hn, ← Real.mul_rpow (r_pos z).le (Nat.cast_nonneg 0)]
+    simp only [Int.cast_zero, zero_mul, add_zero, map_zero, CharP.cast_eq_zero, mul_zero, le_refl]
+  · simp only [ Fin.isValue, Pi.zero_apply, Int.mem_box, Int.natAbs_zero, max_self] at hx
+    have hx2 : x ≠ 0 := by
+      contrapose! hn
+      simp only [hn, Fin.isValue, Pi.zero_apply, Int.natAbs_zero, max_self] at hx
+      exact hx.symm
+    rw [Real.rpow_neg (by apply apply_nonneg), Real.rpow_neg ((r_pos z).le),
+      Real.rpow_neg (Nat.cast_nonneg n), ← mul_inv, inv_le_inv, ← hx, Nat.cast_max]
+    simpa only [Fin.isValue, Nat.cast_max] using (rpow_bound hk z x hx2)
+    · apply Real.rpow_pos_of_pos (Complex.abs.pos (linear_ne_zero ![x 0, x 1] z ?_))
+      have := (Function.comp_ne_zero_iff x Int.cast_injective Int.cast_zero (γ := ℝ)).mpr hx2
+      rw [← Iff.ne (Function.Injective.eq_iff (Equiv.injective (piFinTwoEquiv fun _ ↦ ℝ)))] at this
+      simpa only [Fin.isValue, ne_eq, Matrix.cons_eq_zero_iff, Int.cast_eq_zero, Matrix.zero_empty,
+        and_true, not_and, piFinTwoEquiv_apply, Function.comp_apply, Pi.zero_apply,
+        Prod.mk.injEq] using this
+    · apply mul_pos (Real.rpow_pos_of_pos (r_pos z) _)
+      apply Real.rpow_pos_of_pos (Nat.cast_pos.mpr (Nat.pos_of_ne_zero hn))
```

Comment:
Then this whole proof shortens to
```suggestion
  by_cases hn : n = 0
  · simp only [hn, box_zero, Finset.mem_singleton, Prod.mk_eq_zero] at hx
    rw [hx.1, hx.2, hn, ← Real.mul_rpow (r_pos z).le (Nat.cast_nonneg 0)]
    simp only [Int.cast_zero, zero_mul, add_zero, map_zero, CharP.cast_eq_zero, mul_zero, le_refl]
  · simp only [Fin.isValue, Pi.zero_apply, Int.mem_box, Int.natAbs_zero, max_self] at hx
    have hx2 : x ≠ 0 := by
      contrapose! hn
      simpa only [hn, Fin.isValue, Pi.zero_apply, Int.natAbs_zero, max_self] using hx.symm
    rw [← Real.mul_rpow (r_pos _).le (Nat.cast_nonneg _)]
    exact Real.rpow_le_rpow_of_nonpos (mul_pos (r_pos _) (by positivity)) (hx ▸ r_mul_max_le z hx2) 
      (neg_nonpos.mpr hk)
```

Reply:
oh very nice, I've done this this and used the `r_mul_max_le` name you suggest in your proof :)